### PR TITLE
checkers: add ruleguard checker

### PR DESCRIPTION
**Pull request by quasilyte**
This is part of the ruleguard integration experiment.

Hopefully, we can get some feedback to improve the way it's integrated.

The simplest way to try it out:
```
	gocritic check -enable ruleguard -@ruleguard.rules <gorules> <target>
```

Where:
```
	<gorules> is a path to a Go rules file
	<target> is a linting target
```

An example gorules file can be found here: https://github.com/quasilyte/go-ruleguard/blob/master/rules.go

Visit https://github.com/quasilyte/go-ruleguard for more info on the ruleguard.

Updates #873
Updates #869
Updates #837
Updates #836
Updates #831

Signed-off-by: Iskander Sharipov <quasilyte@gmail.com>

### CODE REVIEW

**Comment by quasilyte** [comment]
I'm going to add tests as well.

**Comment by quasilyte** [comment]
Done. Now we have integration tests for ruleguard.

File: checkers/ruleguard_checker.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+	var info lintpack.CheckerInfo
+	info.Name = "ruleguard"
+	info.Tags = []string{"style", "experimental"}
+	info.Params = lintpack.CheckerParams{
+		"rules": {
+			Value: "",
+			Usage: "path to a gorules file",
+		},
+	}
+	info.Summary = "Detects custom rules violations"
```
Comment (cristaloleg) [inline checkers/ruleguard_checker.go]: Sounds depressive. `Allows users to define custom rules`, WDYT ?

File: checkers/ruleguard_checker.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+type ruleguardChecker struct {
+	ctx *lintpack.CheckerContext
+
+	rset *ruleguard.GoRuleSet
+}
+
+func (c *ruleguardChecker) WalkFile(f *ast.File) {
+	// Rule set is nil if no rules were loaded from a rules file.
```
Comment (cristaloleg) [inline checkers/ruleguard_checker.go]: do we need this? :)

File: checkers/ruleguard_checker.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+func newRuleguardChecker(info *lintpack.CheckerInfo, ctx *lintpack.CheckerContext) *ruleguardChecker {
+	c := &ruleguardChecker{ctx: ctx}
+	rulesFilename := info.Params.String("rules")
+	if rulesFilename == "" {
+		return c
+	}
+
+	// We can handle initialization errors easier when we make
```
Comment (cristaloleg) [inline checkers/ruleguard_checker.go]: // TODO ?

File: checkers/ruleguard_checker.go [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+	data, err := ioutil.ReadFile(rulesFilename)
+	if err != nil {
+		log.Printf("init ruleguard: %v", err)
+		return c
+	}
+
+	fset := token.NewFileSet()
+	rset, err := ruleguard.ParseRules(rulesFilename, fset, bytes.NewReader(data))
+	if err != nil {
+		log.Printf("init ruleguard: %v", err)
```
Comment (cristaloleg) [inline checkers/ruleguard_checker.go]: what about: `ruleguard init error: %#v` ? it's a bit easier to grep then.

File: checkers/ruleguard_checker.go [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+		return
+	}
+
+	ctx := &ruleguard.Context{
+		Pkg:   c.ctx.Pkg,
+		Types: c.ctx.TypesInfo,
+		Sizes: c.ctx.SizesInfo,
+		Fset:  c.ctx.FileSet,
+		Report: func(n ast.Node, msg string, _ *ruleguard.Suggestion) {
+			// TODO(quasilyte): investigate whether we should add a rule name as
```
Comment (cristaloleg) [inline checkers/ruleguard_checker.go]: imo - yes :) but it's a minor one

File: checkers/ruleguard_checker.go (hunk 1 above)
Comment (quasilyte) [reply checkers/ruleguard_checker.go]: Fixed.

File: checkers/ruleguard_checker.go (hunk 3 above)
Comment (quasilyte) [reply checkers/ruleguard_checker.go]: Fixed.

File: checkers/ruleguard_checker.go (hunk 4 above)
Comment (quasilyte) [reply checkers/ruleguard_checker.go]: Fixed.

File: checkers/ruleguard_checker.go (hunk 2 above)
Comment (quasilyte) [reply checkers/ruleguard_checker.go]: Deleted.

**Pull request merged**